### PR TITLE
Add Connector EventListener registration

### PR DESCRIPTION
**Pull request by s2lomon**
It adds a possibility for Connector to define its own event handlers, that will be registered
on connector creation.

### CODE REVIEW

**Comment by electrum** [comment]
Do you have some example use cases in mind?

**Comment by martint** [comment]
@electrum, this is so that plugins that provide connectors can also provide an event listener without having to provide a separate configuration for the event listener. For instance the MySQL connector would be able to provide an event listener that would write to the same MySQL database.

File: presto-tests/src/test/java/io/prestosql/execution/EventsBuilder.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import com.google.common.collect.ImmutableList;
+import io.prestosql.spi.eventlistener.QueryCompletedEvent;
+import io.prestosql.spi.eventlistener.QueryCreatedEvent;
+import io.prestosql.spi.eventlistener.SplitCompletedEvent;
+
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+class EventsBuilder
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/execution/EventsBuilder.java]: Please extract this class in separate commit. Same for `TestingEventListener`

File: presto-tests/src/test/java/io/prestosql/execution/EventsAwaitingQueries.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import org.intellij.lang.annotations.Language;
+
+import java.util.Optional;
+
+import static com.google.common.base.Strings.nullToEmpty;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static org.testng.Assert.fail;
+
+class EventsAwaitingQueries
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/execution/EventsAwaitingQueries.java]: I don't like this class, I would inline it. It looks much simpler to use `AbstractTestFramework` or `QueryAssertions` (preferable) to verify queries, where events could be handled manually. Something like we do `io.prestosql.sql.query.TestRowFilter`.

File: presto-main/src/test/java/io/prestosql/eventlistener/TestConnectorEventListener.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        public EventListener getEventListener()
+        {
+            getEventListenerInvocationCounter.incrementAndGet();
+            return new EventListener()
+            {
+            };
+        }
+    }
+
+    private static class ConnectorFactoryStub
```
[the remark this answers is not in the document]
Comment (s2lomon) [reply presto-main/src/test/java/io/prestosql/eventlistener/TestConnectorEventListener.java]: I can change it to MockConnectorFactory :)

File: presto-tests/src/test/java/io/prestosql/execution/EventsAwaitingQueries.java (hunk 2 above)
Comment (s2lomon) [reply presto-tests/src/test/java/io/prestosql/execution/EventsAwaitingQueries.java]: It's a wrapper around custom written code to execute a query and block until all expected events are gathered. It's extracted from existing tests so it's not really a case for QueryAssertions etc. I could not extract it and add this test to the other tests, but then the setup for it would grow even bigger

File: presto-spi/src/main/java/io/prestosql/spi/connector/Connector.java [hunk 4]
```diff
@@ -155,6 +156,14 @@ default ConnectorAccessControl getAccessControl()
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * @throws UnsupportedOperationException if this connector does not have an EventListener
+     */
+    default EventListener getEventListener()
```
Comment (electrum) [inline presto-spi/src/main/java/io/prestosql/spi/connector/Connector.java]: We could make this `Iterable<EventListener>` and return `emptyList()` by default, which seems cleaner as we aren't arbitrarily restricting to a single listener, and we don't need to deal with `UnsupportedOperationException`.

File: presto-main/src/test/java/io/prestosql/failuredetector/TestHeartbeatFailureDetector.java [hunk 5]
```diff
@@ -1,125 +0,0 @@
...
-
-import static io.airlift.configuration.ConfigBinder.configBinder;
-import static io.airlift.discovery.client.DiscoveryBinder.discoveryBinder;
-import static io.airlift.discovery.client.ServiceTypes.serviceType;
-import static io.airlift.jaxrs.JaxrsBinder.jaxrsBinder;
-import static org.testng.Assert.assertEquals;
-import static org.testng.Assert.assertFalse;
-import static org.testng.Assert.assertTrue;
-
-public class TestHeartbeatFailureDetector
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/failuredetector/TestHeartbeatFailureDetector.java]: This looks like mistake. Undo?

File: presto-main/src/test/java/io/prestosql/connector/MockConnectorFactory.java [hunk 6]
```diff
@@ -350,6 +364,10 @@ public int hashCode()
         private ApplyProjection applyProjection = (session, handle, projections, assignments) -> Optional.empty();
         private BiFunction<ConnectorSession, SchemaTableName, Optional<ConnectorNewTableLayout>> getInsertLayout = defaultGetInsertLayout();
         private BiFunction<ConnectorSession, ConnectorTableMetadata, Optional<ConnectorNewTableLayout>> getNewTableLayout = defaultGetNewTableLayout();
+        private Supplier<EventListener> eventListener = () ->
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/connector/MockConnectorFactory.java]: extract `defaultEventListener`?

File: presto-main/src/test/java/io/prestosql/connector/MockConnectorFactory.java [hunk 7]
```diff
@@ -399,9 +417,25 @@ public Builder withGetNewTableLayout(BiFunction<ConnectorSession, ConnectorTable
             return this;
         }
 
+        public Builder withEventListener(EventListener listener)
+        {
+            requireNonNull(listener, "listener is null");
+
+            this.eventListener = () -> listener;
+            return this;
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/connector/MockConnectorFactory.java]:
```
rnn(eventListener);
return withEventListener(() -> eventListener);
```

File: presto-spi/src/main/java/io/prestosql/spi/connector/Connector.java (hunk 4 above)
Comment (s2lomon) [reply presto-spi/src/main/java/io/prestosql/spi/connector/Connector.java]: Fine with me, that's how it looked like in the first approach.

File: presto-main/src/test/java/io/prestosql/failuredetector/TestHeartbeatFailureDetector.java (hunk 5 above)
Comment (s2lomon) [reply presto-main/src/test/java/io/prestosql/failuredetector/TestHeartbeatFailureDetector.java]: wow that's strange.
